Tests/Realm.Tests/Database/PrimaryKeyTests.cs

```diff
@@ -192,6 +195,70 @@ private RealmObjectBase FindByPKGeneric(Type type, object primaryKeyValue, PKTyp
             return (RealmObjectBase)genericMethod.MakeGenericMethod(type).Invoke(_realm, new[] { primaryKeyValue });
         }
 
+        [Test]
+        public void RealmFind_WhenPKIsIntAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<IntPrimaryKeyWithValueObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsCharAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyCharObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsByteAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyByteObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsInt16AndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyInt16Object>();
+
+        [Test]
+        public void RealmFind_WhenPKIsInt32AndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyInt32Object>();
+
+        [Test]
+        public void RealmFind_WhenPKIsInt64AndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyInt64Object>();
+
+        [Test]
+        public void RealmFind_WhenPKIsStringAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyStringObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsGuidAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyGuidObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsObjectIdAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyObjectIdObject>();
+
+        // In Find<T> anything that is a primitive type (according to Realm's view) is converted to long?.
+        // FindCore<T> converts its input in RealmValue. Which in turn treats long as PritiveValue of type Int.
+        // Without this pass it'd be impossible to match any PK that's of primitive type, since a PK defined as e.g. short would always be
+        // compared against an int
+        private static Type ConvertToRealmValue(Type toConvert)
+        {
+            var nativeTypes = new ReadOnlyCollection<Type>(new Type[] { typeof(int), typeof(long), typeof(short), typeof(byte), typeof(char) });
+
+            foreach (var type in nativeTypes)
+            {
+                if (toConvert == type)
+                {
+                    return typeof(int);
+                }
+            }
+
+            return null;
+        }
+
+        private void RealmFind_IncorrectPKArgument_Throws<T>()
+            where T : RealmObject
+        {
+            var pkInClass = typeof(T).GetProperties().Single(prop => Attribute.IsDefined(prop, typeof(PrimaryKeyAttribute)));
+            var pkType = ConvertToRealmValue(pkInClass.PropertyType) ?? pkInClass.PropertyType;
+
+            foreach (var pk in _primaryKeyValues)
+            {
+                if (pk.GetType() == pkType)
```

Comment:
This check can be folded into a `.Where` clause - I have a modification of Atwood's Law, which goes something like "if something can be rewritten as LINQ, it must be rewritten as LINQ" 😄